core/src/main/java/dk/alexandra/fresco/framework/network/async/AsyncNetwork.java

```diff
@@ -0,0 +1,336 @@
+    for (int i = conf.getMyId() + 1; i <= conf.noOfParties(); i++) {
+      Party p = conf.getParty(i);
+      SocketAddress addr = new InetSocketAddress(p.getHostname(), p.getPort());
+      boolean connectionMade = false;
+      int attempts = 0;
+      while (!connectionMade) {
+        try {
+          SocketChannel channel = SocketChannel.open();
+          channel.connect(addr);
+          channel.configureBlocking(true);
+          this.channelMap.put(i, channel);
+          ByteBuffer b = ByteBuffer.allocate(1);
+          b.put((byte) conf.getMyId());
+          b.position(0);
+          channel.write(b);
+          connectionMade = true;
+          channelMap.put(i, channel);
+          logger.info("P{} connected to {}", conf.getMyId(), p);
+        } catch (IOException e) {
+          attempts++;
+          Thread.sleep(1 << attempts);
+        }
+      }
+    }
+    return channelMap;
+  }
+
+
+  /**
+   * Listens for connections from the opposing parties.
+   * <p>
+   * The resulting connections will be used to receive messages.
+   * </p>
+   *
+   * @throws IOException thrown if an {@link IOException} occurs while listening.
+   */
+  private Map<Integer, SocketChannel> connectServer() throws IOException {
+    Map<Integer, SocketChannel> channelMap = new HashMap<>(conf.getMyId() - 1);
+    for (int i = 1; i < conf.getMyId(); i++) {
+      SocketChannel channel = server.accept();
+      channel.configureBlocking(true);
+      ByteBuffer buf = ByteBuffer.allocate(1);
+      channel.read(buf);
+      buf.position(0);
+      final int id = buf.get();
+      this.channelMap.put(id, channel);
+      logger.info("P{} accepted connection from {}", conf.getMyId(), conf.getParty(id));
+      channelMap.put(id, channel);
+    }
+    server.close();
+    return channelMap;
+  }
+
+  //
+
+  /**
+   * Binds the server to the port of this party.
+   */
+  private void bindServer() {
+    SocketAddress sock = new InetSocketAddress(conf.getMe().getPort());
+    try {
+      this.server = ServerSocketChannel.open();
+      this.server.bind(sock);
+      logger.info("P{} bound at {}", conf.getMyId(), sock);
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to bind to " + sock, e);
+    }
+  }
+
+  @Override
+  public void send(int partyId, byte[] data) {
+    if (!inRange(partyId)) {
+      throw new IllegalArgumentException(
+          "Party id " + partyId + " not in range 1 ... " + getNoOfParties());
+    }
+    if (this.conf.getMyId() == partyId) {
+      this.queues.get(conf.getMyId()).offer(data);
+    } else {
+      this.senderServices.get(partyId).submit(() -> {
+        try {
+          ByteBuffer buf = ByteBuffer.allocate(Integer.BYTES + data.length);
+          buf.putInt(data.length);
+          buf.put(data);
+          buf.position(0);
+          while (buf.hasRemaining()) {
+            this.channelMap.get(partyId).write(buf);
+          }
+        } catch (Exception e) {
+          e.printStackTrace();
```

Comment:
No

Reply:
"No" to printing the exception?

Reply:
Yes, they need to be logged

Reply:
Oh darn. No, exceptions should be logged